src/Filesystem/Dir.php

```diff
@@ -229,131 +229,134 @@ public static function inventory(
-				];
+				$inventory['children'][] = static::inventoryChild($root, $contentExtension, $multilang);
 			} else {
 				$extension = pathinfo($item, PATHINFO_EXTENSION);
 
-				switch ($extension) {
-					case 'htm':
-					case 'html':
-					case 'php':
-						// don't track those files
-						break;
-					case $contentExtension:
-						$content[] = pathinfo($item, PATHINFO_FILENAME);
-						break;
-					default:
-						$inventory['files'][$item] = [
-							'filename'  => $item,
-							'extension' => $extension,
-							'root'      => $root,
-						];
+				// don't track those files
+				if (
+					$extension === 'htm' ||
+					$extension === 'html' ||
+					$extension === 'php'
+				) {
+					continue;
 				}
-			}
-		}
 
-		// remove the language codes from all content filenames
-		if ($multilang === true) {
-			foreach ($content as $key => $filename) {
-				$content[$key] = pathinfo($filename, PATHINFO_FILENAME);
-			}
+				if ($extension === $contentExtension) {
+					$filename = pathinfo($item, PATHINFO_FILENAME);
+
+					// remove the language codes from all content filenames
+					if ($multilang === true) {
+						$filename = pathinfo($filename, PATHINFO_FILENAME);
+					}
 
-			$content = array_unique($content);
+					$content[] = $filename;
+					continue;
+				}
+
+				$inventory['files'][$item] = [
+					'filename'  => $item,
+					'extension' => $extension,
+					'root'      => $root,
+				];
+			}
 		}
 
-		$inventory = static::inventoryContent($inventory, $content);
-		$inventory = static::inventoryModels($inventory, $contentExtension, $multilang);
+		$content 			   = array_unique($content);
+		$inventory['template'] = static::inventoryTemplate(
+			$content,
+			$inventory['files']
+		);
 
 		return $inventory;
 	}
 
 	/**
-	 * Take all content files,
-	 * remove those who are meta files and
-	 * detect the main content file
+	 * Gathers all inventory information for a child,
+	 * incl. matching the correct page model
 	 */
-	protected static function inventoryContent(array $inventory, array $content): array
-	{
-		// filter meta files from the content file
-		if (empty($content) === true) {
-			$inventory['template'] = 'default';
-			return $inventory;
+	protected static function inventoryChild(
+		string $root,
+		string $contentExtension = 'txt',
+		bool $multilang = false
+	): array {
+		$name = pathinfo($root, PATHINFO_BASENAME);
+
+		// extract the slug and num of the directory
+		if (preg_match('/^([0-9]+)' . static::$numSeparator . '(.*)$/', $name, $match)) {
+			$num  = (int)$match[1];
+			$slug = $match[2];
 		}
 
-		foreach ($content as $contentName) {
-			// could be a meta file. i.e. cover.jpg
-			if (isset($inventory['files'][$contentName]) === true) {
-				continue;
+		// adapt the content extension for mult-language setups
+		if ($multilang === true) {
+			$contentExtension = App::instance()->defaultLanguage()->code() . '.' . $contentExtension;
+		}
+
+		// For many models, one call to scandir() is faster
```

Comment:
Is this still true if the page has a lot of subpages?

Reply:
Unclear. I find this really hard to measure and to decide where to focus on performance, it seems to be always a trade off between no of files, templates etc.